plenum/test/consensus/checkpoint_service/test_checkpoint_service_unit.py

```diff
@@ -0,0 +1,241 @@
+    checkpoint_service._bus.subscribe(StartBackupCatchup, catch_msg)
+
+    quorum = checkpoint_service._data.quorums.checkpoint.value
+    print(quorum)
+    n = len(validators)
+    assert quorum == n - getMaxFailures(n) - 1
+    senders = ["sender{}".format(i) for i in range(quorum + 1)]
+
+    old_key = (1, tconf.CHK_FREQ)
+    key = (old_key[1] + 1, old_key[1] + tconf.CHK_FREQ)
+
+    new_checkpoint = Checkpoint(instId=ordered.instId,
+                                viewNo=ordered.viewNo,
+                                seqNoStart=key[0],
+                                seqNoEnd=key[1],
+                                digest='digest')
+
+    for sender in senders[:quorum]:
+        assert not checkpoint_service.process_checkpoint(checkpoint, sender)
+        assert checkpoint_service._stashed_recvd_checkpoints[checkpoint.viewNo][old_key][sender] == checkpoint
+
+    for sender in senders[:quorum - 1]:
+        assert not checkpoint_service.process_checkpoint(new_checkpoint, sender)
+        assert checkpoint_service._stashed_recvd_checkpoints[checkpoint.viewNo][key][sender] == new_checkpoint
+
+    assert not checkpoint_service.process_checkpoint(new_checkpoint, senders[quorum - 1])
+
+    if is_master:
+        assert checkpoint_service._data.low_watermark == key[1]
+        assert isinstance(caught_msg, StartMasterCatchup)
+    else:
+        assert isinstance(caught_msg, StartBackupCatchup)
+        assert caught_msg.caught_up_till_3pc == (checkpoint_service.view_no, key[1])
+
+
+def test_process_backup_catchup_msg(checkpoint_service, tconf, checkpoint):
+    checkpoint_service._data.last_ordered_3pc = (checkpoint_service.view_no, 0)
+    key = (1, tconf.CHK_FREQ)
+    new_key = (key[1] + 1, key[1] + tconf.CHK_FREQ)
+    checkpoint_service._data.stable_checkpoint = 1
+
+    checkpoint_service._stash_checkpoint(Checkpoint(1, checkpoint.viewNo, new_key[0], new_key[1], "1"),
+                                         "frm")
+    checkpoint_service._stash_checkpoint(Checkpoint(1, checkpoint.viewNo, key[0], key[1], "1"),
+                                         "frm")
+    checkpoint_service._checkpoint_state[key] = CheckpointState(key[1] - 1,
+                                                                ["digest"] * (tconf.CHK_FREQ - 1),
+                                                                None,
+                                                                {},
+                                                                False)
+    checkpoint_service._data.checkpoints.append(checkpoint)
+
+    # send StartBackupCatchup to the internal bus
+    checkpoint_service._bus.send(StartBackupCatchup((checkpoint_service.view_no, key[1])))
+
+    assert checkpoint_service._data.low_watermark == key[1]
+    assert not checkpoint_service._checkpoint_state
+    assert not checkpoint_service._data.checkpoints
+    assert checkpoint_service._data.stable_checkpoint == 0
```

Comment:
Why stable checkpoint is reset to 0? As I understand, it was 1.

Reply:
In this case on master we are calling ConsensusDataHelper.reset_checkpoints() which do
```self.consensus_data.stable_checkpoint = 0```
I think this is correct because stable_checkpoint = 1 means that a some checkpoint already stabilized.